powersimdata/input/converter/pypsa_to_grid.py

```diff
@@ -0,0 +1,328 @@
+
+        # branch
+        drop_cols = pypsa_import_const["branch"]["drop_cols_in_advance"]
+        df = n.lines.drop(columns=drop_cols, errors="ignore")
+        lines = self._translate_df(df, "branch")
+        lines["branch_device_type"] = "Line"
+
+        df = n.transformers.drop(columns=drop_cols, errors="ignore")
+        transformers = self._translate_df(df, "branch")
+        if "branch_device_type" not in transformers:
+            transformers["branch_device_type"] = "Transfomer"
+
+        branch = pd.concat([lines, transformers])
+        branch["x"] *= 100
+        branch["r"] *= 100
+        branch["from_bus_id"] = pd.to_numeric(branch.from_bus_id, errors="ignore")
+        branch["to_bus_id"] = pd.to_numeric(branch.to_bus_id, errors="ignore")
+        branch.index.name = "branch_id"
+
+        # DC lines
+        df = n.df("Link")[lambda df: df.index.str[:3] != "sub"]
+        dcline = self._translate_df(df, "link")
+        dcline["Pmin"] *= dcline["Pmax"]  # convert relative to absolute
+        dcline["from_bus_id"] = pd.to_numeric(dcline.from_bus_id, errors="ignore")
+        dcline["to_bus_id"] = pd.to_numeric(dcline.to_bus_id, errors="ignore")
+
+        # storages
+        if not n.storage_units.empty or not n.stores.empty:
+            warnings.warn("The export of storages are not implemented yet.")
+
+        # Drop columns if wanted
+        if drop_cols:
+            self._drop_cols(bus, "bus")
+            self._drop_cols(plant, "generator")
+            self._drop_cols(branch, "branch")
+            self._drop_cols(dcline, "link")
+
+        # Pull operational properties into grid object
+        if len(n.snapshots) == 1:
+            bus = bus.assign(**self._translate_pnl(n.pnl("Bus"), "bus"))
+            bus["Va"] = np.rad2deg(bus["Va"])
+            bus = bus.assign(**self._translate_pnl(n.pnl("Load"), "bus"))
+            plant = plant.assign(**self._translate_pnl(n.pnl("Generator"), "generator"))
+            _ = pd.concat(
+                [
+                    self._translate_pnl(n.pnl(c), "branch")
+                    for c in ["Line", "Transformer"]
+                ]
+            )
+            branch = branch.assign(**_)
+            dcline = dcline.assign(**self._translate_pnl(n.pnl("Link"), "link"))
+        else:
+            plant["status"] = n.generators_t.status.any().astype(int)
+
+        # Convert to numeric
+        for df in (bus, sub, bus2sub, gencost, plant, branch, dcline):
+            df.index = pd.to_numeric(df.index, errors="ignore")
+
+        self.data_loc = data_loc
+        self.interconnect = interconnect
+        self.bus = bus
+        self.sub = sub
+        self.bus2sub = bus2sub
+        self.branch = branch.sort_index()
+        self.dcline = dcline
+        self.zone2id = zone2id
+        self.id2zone = id2zone
+        self.plant = plant
+        self.gencost["before"] = gencost
+        self.gencost["after"] = gencost
+
+    def _drop_cols(self, df, key):
```

Comment:
Looks like the methods from here onward could be changed to functions and exist outside the class.

Reply:
We can move them in the `helpers` module

Reply:
That could work, or keeping them in the same file - either way.

Reply:
I will see if it can be reused elsewhere